docs/DifferentiableProgramming.md

```diff
@@ -1670,13 +1848,122 @@ func _(_ x: Float) -> (value: Float,
 }
 ```
 
-A derivative function has the same parameters as the original function, but
-returns a linear
-[differential](https://en.wikipedia.org/wiki/Pushforward_\(differential\))
-function in addition to the original value. Computing both the original value
-and the differential is the most efficient way for the differential closure to
-capture anything it needs from the original computation, and is important for
-flexibility and performance.
+##### Typing rules
+
+A function declaration does not have a fixed derivative type. This is because
+there can be multiple derivative functions that differentiate the original
+function differently, e.g. differentiating with respect to different parameters,
+differentiating with different generic constraints, etc.
+
+Given an original function declaration, a derivative function's type is
+determined from the following configurations:
+- Parameters to differentiate with respect to, aka. differentiability
+  parameters.
+- Additional generic constraints that make the original function differentiable.
+
+The type of the derivative function under such configurations is a function that
+takes the original function's parameters and returns a tuple of an original
+result (labeled `value`) and a differential (labeled `differential`). The
```

Comment:
Do we have a note, and if not, should we have a note saying we currently only support `pullback:`?

Reply:
I think we can omit `pullback:`. The manifesto describes the final design for differentiable programming, in which users do not write primitive VJP functions with `pullback:` labels.

Reply:
oh true!